ml-agents/mlagents/trainers/models_torch.py

```diff
@@ -142,6 +179,116 @@ def forward(self, vec_inputs, vis_inputs, memories=None, sequence_length=1):
         return embedding, memories
 
 
+class QNetwork(NetworkBody):
+    def __init__(  # pylint: disable=W0231
+        self,
+        stream_names: List[str],
+        vector_sizes: List[int],
+        visual_sizes: List[CameraResolution],
+        network_settings: NetworkSettings,
+        act_type: ActionType,
+        act_size: List[int],
+    ):
+        # This is not a typo, we want to call __init__ of nn.Module
+        nn.Module.__init__(self)
+        self.normalize = network_settings.normalize
+        self.visual_encoders = []
+        self.vector_encoders = []
+        self.vector_normalizers = []
+        self.use_lstm = network_settings.memory is not None
+        self.h_size = network_settings.hidden_units
+        self.m_size = (
+            network_settings.memory.memory_size
+            if network_settings.memory is not None
+            else 0
+        )
+
+        visual_encoder = ModelUtils.get_encoder_for_type(
+            network_settings.vis_encode_type
+        )
+        for vector_size in vector_sizes:
+            if vector_size != 0:
+                self.vector_normalizers.append(Normalizer(vector_size))
+                input_size = (
+                    vector_size + sum(act_size)
+                    if not act_type == ActionType.DISCRETE
+                    else vector_size
+                )
+                self.vector_encoders.append(
+                    VectorEncoder(input_size, self.h_size, network_settings.num_layers)
+                )
+        for visual_size in visual_sizes:
+            self.visual_encoders.append(
+                visual_encoder(
+                    visual_size.height,
+                    visual_size.width,
+                    visual_size.num_channels,
+                    self.h_size,
+                )
+            )
+
+        self.vector_encoders = nn.ModuleList(self.vector_encoders)
+        self.visual_encoders = nn.ModuleList(self.visual_encoders)
+        if self.use_lstm:
+            self.lstm = nn.LSTM(self.h_size, self.m_size // 2, 1)
+        if act_type == ActionType.DISCRETE:
+            self.q_heads = ValueHeads(
+                stream_names, network_settings.hidden_units, sum(act_size)
+            )
+        else:
+            self.q_heads = ValueHeads(stream_names, network_settings.hidden_units)
+
+    def forward(  # pylint: disable=W0221
```

Comment:
Actually, it seems. that there is a lot of duplication of NetworkBody's forward function here.

Reply:
The main reason for this is the normalizers that are built into the NetworkBody - we don't want to normalize the action so I modified the forward to do this (this is also why I wasn't able to make this contain a NetworkBody). Ideally I'd like it to contain a NetworkBody and re-use it's forward, but to do that we need to first separate the Normalizers from the NetworkBody. 

So a class like Critic would have `Normalizers -> NetworkBody -> ValueHeads` and an Actor would have `Normalizers -> NetworkBody -> Distribution`. A Q-Network would then become `Normalizers + Action -> NetworkBody -> ValueHeads`.